api/wasm/cpp/proxy_wasm_intrinsics.cc

```diff
@@ -1,165 +1,193 @@
+    root_factories = new std::unordered_map<std::string, RootFactory>;
+    context_factories = new std::unordered_map<std::string, ContextFactory>;
+  }
+  if (context_factory)
+    (*context_factories)[std::string(root_id)] = context_factory;
+  if (root_factory)
+    (*root_factories)[std::string(root_id)] = root_factory;
+}
+
+static Context* ensureContext(uint32_t context_id, uint32_t root_context_id) {
   auto e = context_map.insert(std::make_pair(context_id, nullptr));
-  if (e.second)
-    e.first->second = NewContext(context_id);
-  return e.first->second.get();
+  if (e.second) {
+    RootContext *root = context_map[root_context_id].get()->asRoot();
+    std::string root_id = std::string(root->root_id());
+    if (!context_factories) {
+      e.first->second = std::make_unique<Context>(context_id, root);
+      return e.first->second->asContext();
+    }
+    auto factory = context_factories->find(root_id);
+    if (factory == context_factories->end()) {
+      throw ProxyException("no context factory for root_id: " + root_id);
+    } else {
+      e.first->second = factory->second(context_id, root);
+    }
+  }
+  return e.first->second->asContext();
+}
+
+static RootContext* ensureRootContext(uint32_t context_id, std::unique_ptr<WasmData> root_id) {
+  auto it = context_map.find(context_id);
+  if (it != context_map.end()) {
+    return it->second->asRoot();
+  }
+  if (!root_factories) {
+    auto context = std::make_unique<RootContext>(context_id, root_id->view());
+    RootContext* root_context = context->asRoot();
+    context_map[context_id] = std::move(context);
+    return root_context;
+  }
+  auto root_id_string = root_id->toString();
+  auto factory = root_factories->find(root_id_string);
+  RootContext *root_context;
+  if (factory != root_factories->end()) {
+    auto context = factory->second(context_id, root_id->view());
+    root_context = context->asRoot();
+    root_context_map[root_id_string] = root_context;
+    context_map[context_id] = std::move(context);
+  } else {
+    // Default handlers.
+    auto context = std::make_unique<RootContext>(context_id, root_id->view());
+    root_context = context->asRoot();
+    context_map[context_id] = std::move(context);
+  }
+  return root_context;
+}
+
+static ContextBase* getContextBase(uint32_t context_id) {
+  auto it = context_map.find(context_id);
+  if (it == context_map.end() || !it->second->asContext()) {
+    throw ProxyException("no base context context_id: " + std::to_string(context_id));
+  }
+  return it->second.get();
 }
 
 static Context* getContext(uint32_t context_id) {
   auto it = context_map.find(context_id);
-  if (it == context_map.end())
-    return nullptr;
-  return it->second.get();
+  if (it == context_map.end() || !it->second->asContext()) {
+    throw ProxyException("no context context_id: " + std::to_string(context_id));
+  }
+  return it->second->asContext();
+}
+
+static RootContext* getRootContext(uint32_t context_id) {
```

Comment:
You would need to have some shared component, e.g. a RootContextXXXBase which is the superclass for all RootContex's shared in your project.  You would need to have the type definitions for whatever you want to share, so there will have to be some shared types in any case, so I don't think this would be too onerous.